lib/IRGen/GenOpaque.cpp

```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
   Builder.CreateCall(stackRestoreFn, savedSP);
 }
 
+llvm::Value *IRGenFunction::emitStackAlloc(llvm::Value *size,
+                                           llvm::Value *align) {
+  auto returnType = IGM.Int8Ty;
+  auto returnPtrType = returnType->getPointerTo();
+  StackAddress stackPointer;
+  auto nullPointer = llvm::Constant::getNullValue(returnPtrType);
+
+  // Check that this function type can emit a stack allocation safely.
+  if (isAsync() || isCoroutine()) {
+    // These types of functions don't actually stack-allocate, and support for
+    // them is not needed at this time.
+    return nullPointer;
+  }
+
+  // Check that the caller-supplied alignment is usable.
+  Alignment resolvedAlign;
+  if (auto alignConst = dyn_cast<llvm::ConstantInt>(align)) {
+    // Ensure the supplied alignment is itself a power of 2.
+    auto alignConstValue = alignConst->getSExtValue();
+    if (!llvm::isPowerOf2_64(alignConstValue)) {
+      alignConstValue = llvm::NextPowerOf2(alignConstValue);
+    }
+
+    if (alignConstValue > 0) {
+      resolvedAlign = Alignment(alignConstValue);
+
+    } else {
+      // The specified alignment is nonsensical (negative, zero, or too big.)
+      return nullPointer;
+    }
+
```

Comment:
Maybe this could be simplified a little?

```
auto alignConstValue = alignConst->getSExtValue();
if (alignConstValue <= 0 /* optionally add too big check */) {
  return nullPointer;
}
resolvedAlign = Alignment(llvm::PowerOf2Ceil(alignConstValue)); 
```

Reply:
> Maybe this could be simplified a little?
> 
> ```
> auto alignConstValue = alignConst->getSExtValue();
> if (alignConstValue <= 0 /* optionally add too big check */) {
>   return nullPointer;
> }
> resolvedAlign = Alignment(llvm::PowerOf2Ceil(alignConstValue)); 
> ```

`PowerOf2Ceil()` rounds to zero on overflow, so I don't think this is necessarily a correct change.

Reply:
`NextPowerOf2` also returns `0` on overflow, so I'm not sure what's different? We should probably add an upper limit check to `if (alignConstValue <= 0 /* optionally add too big check */)` anyways.

Reply:
Sorry, yes, I'm using `NextPowerOf2()` already—which returns `0` on overflow.

Oh, I see what you're saying, `PowerOf2Ceil()` won't turn e.g. `8` into `16`, right?

Reply:
It shouldn't turn 8 into 16.

```
/// Returns the power of two which is greater than or equal to the given value.
/// Essentially, it is a ceil operation across the domain of powers of two.
inline uint64_t PowerOf2Ceil(uint64_t A) {
  if (!A)
    return 0;
  return NextPowerOf2(A - 1);
}
```

Reply:
Alright.